File: nodns-resolver/src/zones.ts
import { SimplePool } from 'nostr-tools/pool';
import type { NostrEvent } from 'nostr-tools/pure';
import type {
  DiscoveredZone,
  ZoneCheckOutcome,
  ZoneInfo,
  ZonePricing,
  ZoneStatusLevel,
} from './types.js';
import { DEFAULT_DOH_ENDPOINT, ZONE_HANDLER_KIND } from './types.js';

const QUERY_MAX_WAIT = 6_000;
const DOH_TIMEOUT_MS = 10_000;

export function parseZoneTxt(txt: string): Record<string, string> {
  const result: Record<string, string> = {};
  for (const part of txt.split(';')) {
    const eq = part.indexOf('=');
    if (eq === -1) continue;
    const key = part.slice(0, eq).trim();
    const val = part.slice(eq + 1).trim();
    if (key) result[key] = val;
  }
  return result;
}

interface DohAnswer {
  name: string;
  type: number;
  TTL: number;
  data: string;
}

interface DohResponse {
  Status: number;
  Answer?: DohAnswer[];
}

function isDohResponse(data: unknown): data is DohResponse {
  return typeof data === 'object' && data !== null && 'Status' in data;
}

export async function fetchDnsTxt(
  zone: string,
  dohEndpoint: string = DEFAULT_DOH_ENDPOINT,
): Promise<string | null> {
  const name = `_nodns.${zone}`;
  const url = `${dohEndpoint}?name=${encodeURIComponent(name)}&type=TXT`;
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), DOH_TIMEOUT_MS);
  try {
    const res = await fetch(url, {
      headers: { accept: 'application/dns-json' },
      signal: controller.signal,
    });
    if (!res.ok) return null;
    const data: unknown = await res.json();
    if (!isDohResponse(data) || !data.Answer) return null;
    for (const answer of data.Answer) {
      if (answer.type === 16 && typeof answer.data === 'string') {
        return answer.data
          .replace(/^"/, '')
          .replace(/"$/, '')
          .replace(/"\s*"/g, '');
      }
    }
    return null;
  } catch {
    return null;
  } finally {
    clearTimeout(timeout);
  }
}

function parsePricingTag(
  tags: string[][],
): ZonePricing | undefined {
  const tag = tags.find((t) => t[0] === 'pricing');
  if (!tag) return undefined;
  let create = 0;
  let update = 0;
  let del = 0;
  let found = false;
  for (const entry of tag.slice(1)) {
    const eq = entry.indexOf('=');
    if (eq === -1) continue;
    const key = entry.slice(0, eq).trim();
    const val = parseInt(entry.slice(eq + 1).trim(), 10);
    if (Number.isNaN(val)) continue;
    found = true;
    if (key === 'create') create = val;
    else if (key === 'update') update = val;
    else if (key === 'delete') del = val;
  }
  return found ? { create, update, delete: del } : undefined;
}

function parseStatusTag(
  tags: string[][],
): { status: ZoneStatusLevel; reason?: string } {
  const tag = tags.find((t) => t[0] === 'status' && t[1]);
  if (!tag || !tag[1]) return { status: 'unknown' };
  const value = tag[1].toLowerCase();
  if (value === 'testing' || value === 'preview' || value === 'production') {
    return { status: value, reason: tag[2] };
  }
  return { status: 'unknown' };
}

function eventToPartialZone(event: NostrEvent): DiscoveredZone | null {
  const tags = event.tags;
  const zoneTag = tags.find((t) => t[0] === 'zone' && t[1]);
  if (!zoneTag) return null;
  const zone = zoneTag[1].toLowerCase();
  const pricing = parsePricingTag(tags);
  const { status, reason } = parseStatusTag(tags);
  const testnet = tags.some((t) => t[0] === 'testnet');
  const dnskeyHashTag = tags.find((t) => t[0] === 'dnskey_hash' && t[1]);
  const dnskeyAlgTag = tags.find((t) => t[0] === 'dnskey_alg' && t[1]);
  const mintTag = tags.find((t) => t[0] === 'mint' && t[1]);
  const webTag = tags.find((t) => t[0] === 'web' && t[1]);
  return {
    zone,
    pubkey: event.pubkey,
    status,
    testnet,
    statusReason: reason,
    dnskeyHash: dnskeyHashTag?.[1],
    dnskeyAlg: dnskeyAlgTag?.[1],
    pricing,
    mint: mintTag?.[1],
    web: webTag?.[1],
    verified: false,
  };
}

async function queryZoneHandlerEvents(
  relays: string[],
): Promise<NostrEvent[]> {
  const pool = new SimplePool();
  try {
    const tagged = await pool.querySync(
      relays,
      { kinds: [ZONE_HANDLER_KIND], '#k': ['11111'] },
      { maxWait: QUERY_MAX_WAIT },
    );
    if (tagged.length > 0) return tagged as NostrEvent[];
    return await pool.querySync(
      relays,
      { kinds: [ZONE_HANDLER_KIND] },
      { maxWait: QUERY_MAX_WAIT },
    ) as NostrEvent[];
  } finally {
    pool.close(relays);
  }
}

export async function discoverZones(
  relays: string[],
  dohEndpoint: string = DEFAULT_DOH_ENDPOINT,
): Promise<DiscoveredZone[]> {
  let events: NostrEvent[];
  try {
    events = await queryZoneHandlerEvents(relays);
  } catch {
    return [];
  }

  const byZone = new Map<string, { event: NostrEvent; ts: number }>();
  for (const event of events) {
    const kTags = event.tags.filter((t) => t[0] === 'k').map((t) => t[1]);
    if (kTags.length > 0 && !kTags.includes('11111')) continue;
    const zoneTag = event.tags.find((t) => t[0] === 'zone' && t[1]);
    if (!zoneTag) continue;
    const zone = zoneTag[1].toLowerCase();
    const existing = byZone.get(zone);
    if (!existing || event.created_at > existing.ts) {
      byZone.set(zone, { event, ts: event.created_at });
    }
  }

  const zones: DiscoveredZone[] = [];
  for (const { event } of byZone.values()) {
    const partial = eventToPartialZone(event);
    if (partial) zones.push(partial);
  }

  await Promise.allSettled(
    zones.map(async (zone) => {
      const txt = await fetchDnsTxt(zone.zone, dohEndpoint);
      if (!txt) {
        zone.verificationError = 'No _nodns TXT record found';
        return;
      }
      const parsed = parseZoneTxt(txt);
      const txtNpub = parsed['npub'];
      if (!txtNpub) {
        zone.verificationError = 'TXT record missing npub field';
        return;
      }
      if (txtNpub.toLowerCase() !== zone.pubkey.toLowerCase()) {
        zone.verificationError = 'TXT npub does not match event signer';
        return;
      }
      if (parsed['testnet'] === '1') {
        zone.testnet = true;
      }
      zone.verified = true;
    }),
  );

  zones.sort((a, b) => a.zone.localeCompare(b.zone));
  return zones;
}

function buildZoneInfo(
  parsed: Record<string, string>,
  zone: string,
): ZoneInfo {
  const npub = parsed['npub'] ?? '';
  const testnet = parsed['testnet'] === '1' || parsed['testnet'] === 'true';
  const npubNamesFree =
    parsed['npub_free'] === undefined
      ? true
      : parsed['npub_free'] === 'true' || parsed['npub_free'] === '1';

  const createPrice = parseInt(parsed['create'] ?? '0', 10) || 0;
  const updatePrice = parseInt(parsed['update'] ?? '0', 10) || 0;
  const deletePrice = parseInt(parsed['delete'] ?? '0', 10) || 0;

  const pricing: ZonePricing | undefined =
    createPrice > 0 || updatePrice > 0 || deletePrice > 0
      ? { create: createPrice, update: updatePrice, delete: deletePrice }
      : undefined;

  let mintUrl = parsed['mint'];
  if (mintUrl && !mintUrl.startsWith('http')) {
    mintUrl = `https://${mintUrl}`;
  }

  return {
    zone,
    npub,
    testnet,
    optedIn: true,
    pricing,
    mintUrl,
    npubNamesFree,
    handlerEventFound: false,
    verified: false,
  };
}

async function fetchZoneHandler(
  relays: string[],
  pubkeyHex: string,
): Promise<NostrEvent | null> {
  const pool = new SimplePool();
  try {
    const events = await pool.querySync(
      relays,
      { kinds: [ZONE_HANDLER_KIND], authors: [pubkeyHex], limit: 1 },
      { maxWait: QUERY_MAX_WAIT },
    );
    return events.length > 0 ? (events[0] as NostrEvent) : null;
  } catch {
    return null;
  } finally {
    pool.close(relays);
  }
}

export async function checkZone(
  zone: string,
  relays: string[],
  opts?: { skipCheck?: boolean; dohEndpoint?: string },
): Promise<ZoneCheckOutcome> {
  const dohEndpoint = opts?.dohEndpoint ?? DEFAULT_DOH_ENDPOINT;

  if (opts?.skipCheck) {
    return {
      result: 'verified',
      info: {
        zone,
        npub: '',
        testnet: false,
        optedIn: true,
        npubNamesFree: true,
        handlerEventFound: true,
        verified: true,
      },
    };
  }

  const txt = await fetchDnsTxt(zone, dohEndpoint);
  if (!txt) {
    return { result: 'not-opted-in' };
  }

  const parsed = parseZoneTxt(txt);
  const info = buildZoneInfo(parsed, zone);

  if (!info.npub) {
    return { result: 'unverified', info };
  }

  const handler = await fetchZoneHandler(relays, info.npub);
  if (handler && handler.pubkey.toLowerCase() === info.npub.toLowerCase()) {
    info.handlerEventFound = true;
    info.verified = true;
    info.optedIn = true;
    if (info.testnet) {
      return { result: 'testnet', info };
    }
    return { result: 'verified', info };
  }

  info.handlerEventFound = handler !== null;
  info.verified = false;
  return { result: 'unverified', info };
}

File: nodns-resolver-rs/src/zones.rs
//! Zone discovery — find nodns zones via Nostr handler events and DNS TXT verification.

use std::collections::HashMap;
use std::time::Duration;

use anyhow::Result;
use nostr_sdk::{Client, Event, Filter, Kind};

use crate::dns::query_doh;
use crate::types::{ZonePricing, ZoneStatus, ZoneStatusLevel, ZoneTxtFields, ZONE_HANDLER_KIND};

const QUERY_TIMEOUT_SECS: u64 = 6;

#[must_use]
pub fn parse_zone_txt(txt: &str) -> ZoneTxtFields {
    let mut result = HashMap::new();
    for part in txt.split(';') {
        if let Some(eq) = part.find('=') {
            let key = part[..eq].trim().to_string();
            let val = part[eq + 1..].trim().to_string();
            if !key.is_empty() {
                result.insert(key, val);
            }
        }
    }
    result
}

fn strip_txt_quotes(data: &str) -> String {
    data.trim_start_matches('"')
        .trim_end_matches('"')
        .replace("\" \"", "")
        .replace("\"", "")
}

pub async fn fetch_dns_txt(
    zone: &str,
    doh_endpoint: &str,
    http: &reqwest::Client,
) -> Option<String> {
    let name = format!("_nodns.{zone}");
    match query_doh(&name, "TXT", doh_endpoint, http).await {
        Ok(resp) => {
            if let Some(answers) = resp.answer {
                for a in answers {
                    if a.type_num == 16 {
                        return Some(strip_txt_quotes(&a.data));
                    }
                }
            }
            None
        }
        Err(_) => None,
    }
}

fn parse_pricing_tag(tags_raw: &[Vec<String>]) -> Option<ZonePricing> {
    let tag = tags_raw
        .iter()
        .find(|t| !t.is_empty() && t[0] == "pricing")?;
    let mut create = 0u64;
    let mut update = 0u64;
    let mut delete = 0u64;
    let mut found = false;

    for entry in &tag[1..] {
        if let Some(eq) = entry.find('=') {
            let key = entry[..eq].trim();
            let val_str = entry[eq + 1..].trim();
            if let Ok(val) = val_str.parse::<u64>() {
                found = true;
                match key {
                    "create" => create = val,
                    "update" => update = val,
                    "delete" => delete = val,
                    _ => {}
                }
            }
        }
    }

    if found {
        Some(ZonePricing {
            create,
            update,
            delete,
        })
    } else {
        None
    }
}

fn parse_status_tag(tags_raw: &[Vec<String>]) -> (ZoneStatusLevel, Option<String>) {
    for t in tags_raw {
        if t.len() >= 2 && !t.is_empty() && t[0] == "status" {
            let value = t[1].to_lowercase();
            let reason = t.get(2).cloned();
            return match value.as_str() {
                "testing" => (ZoneStatusLevel::Testing, reason),
                "preview" => (ZoneStatusLevel::Preview, reason),
                "production" => (ZoneStatusLevel::Production, reason),
                _ => (ZoneStatusLevel::Unknown, reason),
            };
        }
    }
    (ZoneStatusLevel::Unknown, None)
}

fn event_to_partial_zone(event: &Event) -> Option<ZoneStatus> {
    let tags_raw: Vec<Vec<String>> = event.tags.iter().map(|t| t.as_slice().to_vec()).collect();

    let zone_tag = tags_raw.iter().find(|t| t.len() >= 2 && t[0] == "zone")?;
    let zone = zone_tag[1].to_lowercase();

    let pricing = parse_pricing_tag(&tags_raw);
    let (status, status_reason) = parse_status_tag(&tags_raw);

    let testnet = tags_raw.iter().any(|t| !t.is_empty() && t[0] == "testnet");

    let dnskey_hash = tags_raw
        .iter()
        .find(|t| t.len() >= 2 && t[0] == "dnskey_hash")
        .map(|t| t[1].clone());

    let dnskey_alg = tags_raw
        .iter()
        .find(|t| t.len() >= 2 && t[0] == "dnskey_alg")
        .map(|t| t[1].clone());

    let mint = tags_raw
        .iter()
        .find(|t| t.len() >= 2 && t[0] == "mint")
        .map(|t| t[1].clone());

    let web = tags_raw
        .iter()
        .find(|t| t.len() >= 2 && t[0] == "web")
        .map(|t| t[1].clone());

    Some(ZoneStatus {
        zone,
        pubkey: event.pubkey.to_hex(),
        status,
        testnet,
        status_reason,
        dnskey_hash,
        dnskey_alg,
        pricing,
        mint,
        web,
        verified: false,
        verification_error: None,
    })
}

async fn create_client(relays: &[String]) -> Result<Client> {
    let client = Client::builder().build();
    for url in relays {
        client.add_relay(url).await?;
    }
    client.connect().await;
    Ok(client)
}

async fn query_zone_handler_events(relays: &[String]) -> Vec<Event> {
    let client = match create_client(relays).await {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };

    let filter = Filter::new()
        .kind(Kind::Custom(ZONE_HANDLER_KIND as u16))
        .limit(100);

    let events = client
        .fetch_events(filter, Duration::from_secs(QUERY_TIMEOUT_SECS))
        .await
        .map(|e| e.into_iter().collect::<Vec<_>>())
        .unwrap_or_default();

    client.disconnect().await;
    events
}

pub async fn discover_zones(
    relays: &[String],
    doh_endpoint: &str,
    http: &reqwest::Client,
) -> Vec<ZoneStatus> {
    let events = query_zone_handler_events(relays).await;

    let mut by_zone: HashMap<String, (Event, u64)> = HashMap::new();

    for event in &events {
        let tags_raw: Vec<Vec<String>> = event.tags.iter().map(|t| t.as_slice().to_vec()).collect();

        let k_tags: Vec<&str> = tags_raw
            .iter()
            .filter(|t| !t.is_empty() && t[0] == "k")
            .filter_map(|t| t.get(1).map(|s| s.as_str()))
            .collect();

        if !k_tags.is_empty() && !k_tags.contains(&"11111") {
            continue;
        }

        let zone_tag = tags_raw.iter().find(|t| t.len() >= 2 && t[0] == "zone");
        let Some(zone_tag) = zone_tag else { continue };
        let zone = zone_tag[1].to_lowercase();
        let ts = event.created_at.as_secs();

        match by_zone.get(&zone) {
            Some(existing) if ts <= existing.1 => {}
            _ => {
                by_zone.insert(zone, (event.clone(), ts));
            }
        }
    }

    let mut zones: Vec<ZoneStatus> = by_zone
        .values()
        .filter_map(|(event, _)| event_to_partial_zone(event))
        .collect();

    for zone in &mut zones {
        let txt = fetch_dns_txt(&zone.zone, doh_endpoint, http).await;

        let Some(txt) = txt else {
            zone.verification_error = Some("No _nodns TXT record found".to_string());
            continue;
        };

        let parsed = parse_zone_txt(&txt);
        let Some(txt_npub) = parsed.get("npub") else {
            zone.verification_error = Some("TXT record missing npub field".to_string());
            continue;
        };

        if txt_npub.to_lowercase() != zone.pubkey.to_lowercase() {
            zone.verification_error = Some("TXT npub does not match event signer".to_string());
            continue;
        }

        if parsed.get("testnet").map(|s| s.as_str()) == Some("1") {
            zone.testnet = true;
        }

        zone.verified = true;
    }

    zones.sort_by(|a, b| a.zone.cmp(&b.zone));
    zones
}
